var/spack/repos/builtin/packages/exago/package.py

```diff
@@ -178,7 +181,20 @@ class Exago(CMakePackage, CudaPackage, ROCmPackage):
         depends_on("umpire {0}".format(rocm_dep), when="+raja {0}".format(rocm_dep))
         depends_on("camp {0}".format(rocm_dep), when="+raja {0}".format(rocm_dep))
 
-    patch("exago-1.6.0.patch", when="@1.6.0")
+    # CMake patches to support ~python and ~testing
+    for ver in [
+        "1.6.0",
+        "1.5.1",
+        "1.5.0",
+        "1.4.1",
+        "1.4.0",
+        "1.3.0",
+        "1.2.0",
+        "1.1.2",
+        "1.1.1",
+        "1.1.0",
+    ]:
+        patch("exago-{0}.patch".format(ver), when="@{0}".format(ver))
```

Comment:
Probably good to remove duplicate patches from here.

```suggestion
    # CMake patches to support ~python and ~testing
    patch("exago-1.6.0.patch", when="@1.6.0")
    patch("exago-1.5.0.patch", when="@1.5.0:1.5.1")
    patch("exago-1.3.0.patch", when="@1.3.0:1.4.1")
    patch("exago-1.1.0.patch", when="@1.1.0:1.2.0")
```

using `exago-1.5.0.patch` with `exago@1.5.1` works for me.
```
==> Installing exago-1.5.1-tifh5k45ktj56fnib52gf5udrc2qqzh6 [39/39]
==> No binary for exago-1.5.1-tifh5k45ktj56fnib52gf5udrc2qqzh6 found: installing from source
==> Applied patch /home/balay/spack/var/spack/repos/builtin/packages/exago/exago-1.5.0.patch
==> exago: Executing phase: 'cmake'
==> exago: Executing phase: 'build'
==> exago: Executing phase: 'install'
==> exago: Successfully installed exago-1.5.1-tifh5k45ktj56fnib52gf5udrc2qqzh6
  Stage: 1m 8.27s.  Cmake: 2.87s.  Build: 0.86s.  Install: 0.26s.  Post-install: 0.05s.  Total: 1m 12.54s
[+] /home/balay/spack/opt/spack/linux-fedora39-skylake/gcc-13.2.1/exago-1.5.1-tifh5k45ktj56fnib52gf5udrc2qqzh6
```
using `exago-1.3.0.patch` with `exago@1.4.1` also  works - but has subsequent build errors.

```

==> Installing exago-1.4.1-7vpnhzzxsrth7ahzm4pwdvgvpgybcdtf [34/34]
==> No binary for exago-1.4.1-7vpnhzzxsrth7ahzm4pwdvgvpgybcdtf found: installing from source
==> Applied patch /home/balay/spack/var/spack/repos/builtin/packages/exago/exago-1.3.0.patch
==> exago: Executing phase: 'cmake'
==> exago: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j12'

6 errors found in build log:
     434    [ 78%] Linking CXX executable tcopflow
     435    cd /home/balay/spack/spack-stage/spack-stage-exago-1.4.1-7vpnhzzxsrth7ahzm4pwdvgvpgybcdtf/spack-build-7vpnhzz/applications && /usr/bin/cmake -E cmake_link_script CMakeFiles/app_tcopflow.dir/link.txt --verbose=1
     436    /software/mpich-4.1.1/bin/mpic++ -O3 -DNDEBUG CMakeFiles/app_tcopflow.dir/tcopflow_main.cpp.o -o tcopflow  -Wl,-rpath,/home/balay/spack/opt/spack/linux-fedora39-skylake/gcc-13.2.1/petsc-3.16.6-diy2horf5ykuk5kbohcixnvkkxegfehn
            /lib:/home/balay/spack/opt/spack/linux-fedora39-skylake/gcc-13.2.1/netlib-lapack-3.11.0-mbofv6x6fzhx6yragjt3crpcjhrwocp7/lib64:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
            :::::::::::: ../src/tcopflow/libexago_tcopflow.a ../src/opflow/libexago_opflow.a /usr/lib/gcc/x86_64-redhat-linux/13/libgomp.so /usr/lib64/libpthread.a /home/balay/spack/opt/spack/linux-fedora39-skylake/gcc-13.2.1/hiop-1.0.1-
            i7dp5ulk4wfb5iqhmgjab7wowocdhszr/lib/libhiop.a ../src/pflow/libexago_pflow.a ../src/ps/libexago_ps.a ../src/utils/libexago_utils.a ../tpl/spdlog/libspdlog.a /home/balay/spack/opt/spack/linux-fedora39-skylake/gcc-13.2.1/petsc-
            3.16.6-diy2horf5ykuk5kbohcixnvkkxegfehn/lib/libpetsc.so /home/balay/spack/opt/spack/linux-fedora39-skylake/gcc-13.2.1/netlib-lapack-3.11.0-mbofv6x6fzhx6yragjt3crpcjhrwocp7/lib64/liblapack.so /home/balay/spack/opt/spack/linux-
            fedora39-skylake/gcc-13.2.1/netlib-lapack-3.11.0-mbofv6x6fzhx6yragjt3crpcjhrwocp7/lib64/libblas.so
     437    make[2]: Leaving directory '/home/balay/spack/spack-stage/spack-stage-exago-1.4.1-7vpnhzzxsrth7ahzm4pwdvgvpgybcdtf/spack-build-7vpnhzz'
     438    [ 78%] Built target app_tcopflow
     439    In file included from /home/balay/spack/spack-stage/spack-stage-exago-1.4.1-7vpnhzzxsrth7ahzm4pwdvgvpgybcdtf/spack-src/src/scopflow/solver/hiop/scopflow_hiop.cpp:5:
  >> 440    /home/balay/spack/spack-stage/spack-stage-exago-1.4.1-7vpnhzzxsrth7ahzm4pwdvgvpgybcdtf/spack-src/src/scopflow/solver/hiop/scopflow_hiop.h:34:10: error: conflicting return type specified for 'virtual size_t SCOPFLOWHIOPInterfa
            ce::get_num_rterms() const'
     441       34 |   size_t get_num_rterms() const;
     442          |          ^~~~~~~~~~~~~~
```

Reply:
so I re-tried a build of all versions [with the current PR changes]

```console
balay@petsc-gpu-02:/scratch/balay/spack.z$ ./bin/spack -j64 install exago@1.6.0 exago@1.5.1 exago@1.5.0 exago@1.4.1 exago@1.4.0 exago@1.3.0 exago@1.2.0 exago@1.1.2 exago@1.1.1 exago@1.1.0 |& tee build.log
```
The following version built fine
```console
balay@petsc-gpu-02:/scratch/balay/spack.z$ ./bin/spack find exago
-- linux-ubuntu22.04-zen4 / gcc@11.4.0 --------------------------
exago@1.1.0  exago@1.1.1  exago@1.1.2  exago@1.5.0  exago@1.5.1  exago@1.6.0
==> 6 installed packages
```

And others failed.  
[build.log.txt](https://github.com/spack/spack/files/13530664/build.log.txt)


BTW: With my above suggested change [to 4 patch files] - the same versions are build (and with similar build errors for the other versions)

```
balay@petsc-gpu-02:/scratch/balay/spack.zz$ ./bin/spack find exago
-- linux-ubuntu22.04-zen4 / gcc@11.4.0 --------------------------
exago@1.1.0  exago@1.1.1  exago@1.1.2  exago@1.5.0  exago@1.5.1  exago@1.6.0
==> 6 installed packages
```

[build2.log.txt](https://github.com/spack/spack/files/13530761/build2.log.txt)

Reply:
I pushed the above change to remove duplicate patchfiles.

Perhaps `exago-1.6.0.patch` is a commit in the repo? If so - it can be changed to URL from patch.

I'm not sure if `exago@1.4.1 exago@1.4.0 exago@1.3.0 exago@1.2.0` failures should hold up this PR..

Reply:
will go ahead and merge this in - so that 1.5.0 fix is in.

And I guess `exago@1.4.1 exago@1.4.0 exago@1.3.0 exago@1.2.0` fixes - if any can be in a subsequent PR.

Reply:
@balay did you create an issue for those exago build failures / can you make one in ExaGO?

I don't quite understand why you are seeing that error, but that would suggest we have the wrong HiOp versions listed as dependencies for those ExaGO versions, or we need to have a patch in ExaGO / HiOp to address build failures in those versions.

Either way thank you for merging, and apologies for the slow reply as I was taking some time off work.

Reply:
@cameronrutherford, you can grab the logs from here [and steps to reproduce] - and file in your issue tracker [as needed]

I don't need fixes for those versions though..

Reply:
https://github.com/pnnl/ExaGO/issues/101